Convert `BaseFlutterTask` From Groovy to Kotlin

jesswrd opened the pull request:
Converted `BaseFlutterTask` from Groovy to Kotlin.

Note: Since it looks like Gradle prevents us from instantiating and using an object that extends abstract class BaseFlutterTask, we must use principles of Test Driven Development (TDD) in order to unit test `BaseFlutterTask`. `BaseFlutterTaskHelper` was created solely for unit testing to simulate the behavior of all functions in the original `BaseFlutterTask`.  

Fixes #162111 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

jesswrd committed:
hope this passes on CI

# Conflicts:
#	packages/flutter_tools/gradle/src/main/groovy/flutter.groovy

jesswrd committed:
hope this passes on CI

# Conflicts:
#	packages/flutter_tools/gradle/src/main/groovy/flutter.groovy

jesswrd committed:
instantiate buildBundle help

jesswrd committed:
instantiate buildBundle help

jesswrd committed:
added a helper class for testing

jesswrd committed:
added a helper class for testing

jesswrd committed:
capture unit test attempt

jesswrd committed:
capture unit test attempt

jesswrd committed:
things are working

jesswrd committed:
things are working

jesswrd committed:
one working test

jesswrd committed:
one working test

jesswrd committed:
all unit tests working

jesswrd committed:
all unit tests working

jesswrd committed:
removed some comments

jesswrd committed:
removed some comments

jesswrd committed:
debugging

jesswrd committed:
debugging

jesswrd committed:
it buildssss

jesswrd committed:
it buildssss

jesswrd committed:
completed tests

jesswrd committed:
completed tests

jesswrd committed:
Merge branch 'master' into i162111-convert-BaseFlutterTask

jesswrd commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTask.kt:
```diff
@@ -0,0 +1,144 @@
+    @Input
+    var localEngineSrcPath: String? = null
+
+    @Optional
+    @Input
+    var fastStart: Boolean? = null
```
@gmackall It looks like `Boolean?` is boxed just like `boolean` in Java, so it seems like using @Optional here is okay.

reidbaker commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTask.kt:
```diff
@@ -0,0 +1,144 @@
+    @Internal
+    var flutterExecutable: File? = null
+
+    @Input
```
For my own education what does @input do here?

jesswrd replied:
I think @input (Gradle doc [here](https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api.tasks/-input/index.html)) allows Gradle to track the input to the task. I think it's there for efficiency purposes? If it's up-to-date, meaning the property with the @input tag has not changed, then Gradle will skip the associated task. If it's out-of-date, meaning the property with the @input tag has changed, then Gradle will rerun the associated task.

reidbaker replied:
OK and you kept the input annotations from the previous class. That makes sense.

reidbaker commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTask.kt:
```diff
@@ -0,0 +1,144 @@
+
+    @Optional
+    @Input
+    var performanceMeasurementFile: String? = null
```
I am trying to understand which of these values could get moved to BaseFlutterTaskHelper since there are a bunch of variable in this class that are only used in that class. 

So picking on performanceMeasurementFile I see [these places](https://github.com/search?q=repo%3Aflutter%2Fflutter%20performanceMeasurementFile&type=code) it is used. 

Of those one is in [flutter.groovy](https://github.com/flutter/flutter/blob/7e2b4e8f86d53d877b5fd8300094e6cf0f012bb2/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L1281) 

That code is extracting values then setting those values on a `FlutterTask`. 

```
String performanceMeasurementFileValue
final String propPerformanceMeasurementFile = "performance-measurement-file"
if (project.hasProperty(propPerformanceMeasurementFile)) {
      performanceMeasurementFileValue = project.property(propPerformanceMeasurementFile)
}
...
TaskProvider<FlutterTask> compileTaskProvider = project.tasks.register(taskName , FlutterTask) {
        ... 
        performanceMeasurementFile(performanceMeasurementFileValue)
        ... 
}
```

I do not like the broad use of the project object because it is so complex but even in this pr we are using baseFlutterTask.project.<STUFF>. I am open to other opinions but I think it would be worthwhile to have a constructor take in a `project` and pull all of this project evaluation logic out of flutter.groovy and into either basefluttertask or basefluttertaskhelper.

jesswrd replied:
For my own understanding, you want to refactor the usage of properties that are only used in flutter.groovy (or very few files) into basefluttertask or basefluttertaskhelper to group the usages together, so it would be clearer to see where the property is being defined and what task it is being used for. Is that correct?

Also, if we decide to refactor it this way, I think it should be a part of a different issue and potentially done after the groovy to kotlin conversion. It doesn't seem like this refactor has to be included to address this particular issue.

reidbaker replied:
That description is close but there is another element that moving the code that pulls values from the project and sets them on the task. That is scope/context. My moving the code my gut is that many/most variables would not need to exist. ~Part of the challenge at this point is that any variables annotated with @Input would still need to stay since that is what is used for task caching.~ I think we can @input to the project value to make sure we get updated variables but that would mean less efficient caching. 

"If we decide to refactor it this way, I think it should be a part of a different issue and potentially done after the groovy to kotlin conversion." That is not an unreasonable take. I think it is part of the kotlin conversion because the goal of the kotlin conversion is to migrate code out of groovy. That is why in my [pr for AppLinkSettings](https://github.com/flutter/flutter/pull/164391) I also added the ability to generate json to the kotlin classes.  What I am suggesting is aligned with that goal. Also the work of actually converting flutter.groovy will be easier if we can move larger chunks of that work as part of the other conversions. Additionally the ability to test how we pull values from the project object gives us more safety than leaving it in groovy. 

If we decided we didnt want to do this work I think it would be because of another goal of the conversion which is to not "clean up/refactor" how the existing code works? That said in this case if we are getting the data from the same place and preserving the values that can change with input I think it is strictly better. 

@gmackall what do you think about moving some of the values in this task to project accessors.

gmackall replied:
Spoke with @reidbaker offline, I think it would be reasonable to pull out all of the logic in [`addFlutterTasks()`](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L1136) in a separate PR (into its own Kotlin class). This gets all of the logic out of `flutter.groovy`, while still avoiding changing the code behavior.

jesswrd replied:
done. Added the issue [here](https://github.com/flutter/flutter/issues/164996) to the umbrella issue [here](https://github.com/flutter/flutter/issues/162103).

jesswrd commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTaskHelper.kt:
```diff
@@ -0,0 +1,121 @@
+        var depfiles: FileCollection = baseFlutterTask.project.files()
+
+        // Includes all sources used in the flutter compilation.
+        depfiles += baseFlutterTask.project.files("${baseFlutterTask.intermediateDir}/flutter_build.d")
```
From my own search it looks like `flutter_build.d` is just a file with dependencies info in it. I did not find anything online or in my test flutter app that suggests otherwise. But I imagine it's written this way because it could be a directory? I am unsure.

jesswrd committed:
updated intermediateDir to be nonnullable and added ktdocs

jesswrd committed:
cleanup and created baseFlutterTaskHelperTest

ash-google commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTaskHelper.kt:
```diff
@@ -0,0 +1,121 @@
+    // cache.
+    @VisibleForTesting
+    internal fun generateRuleNames(baseFlutterTask: BaseFlutterTask): List<String> {
+        val ruleNames: List<String> =
+            when {
+                baseFlutterTask.buildMode == "debug" -> listOf("debug_android_application")
+                baseFlutterTask.deferredComponents!! ->
+                    baseFlutterTask.targetPlatformValues!!.map {
+                        "android_aot_deferred_components_bundle_${baseFlutterTask.buildMode}_$it"
+                    }
+                else -> baseFlutterTask.targetPlatformValues!!.map { "android_aot_bundle_${baseFlutterTask.buildMode}_$it" }
+            }
+        return ruleNames
```
Let's wrap this into a `with (baseFlutterTask) { ... }` block, get and use more null safe operations instead of forcing each variable to be nonnullable with `!!`.

So it looks more like this:
```
    internal fun generateRuleNames(baseFlutterTask: BaseFlutterTask): List<String> =
        with(baseFlutterTask) {
            when {
                buildMode == "debug" -> listOf("debug_android_application")
                deferredComponents == true ->
                    targetPlatformValues?.map {
                        "android_aot_deferred_components_bundle_${buildMode}_$it"
                    } ?: emptyList()
                else ->
                    targetPlatformValues?.map { "android_aot_bundle_${buildMode}_$it" }
                        ?: emptyList()
            }
        }
```

jesswrd replied:
Since we want an exact Groovy to Kotlin conversion, we will handle NPE handing during cleanup in a different PR.

ash-google commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTaskHelper.kt:
```diff
@@ -0,0 +1,121 @@
+    @VisibleForTesting
+    internal fun createExecSpecActionFromTask(): Action<ExecSpec> =
+        Action<ExecSpec> {
+            executable(baseFlutterTask.flutterExecutable!!.absolutePath)
```
There's a couple of places I see where we're doing these not null assertions. Could we make them a little safer? For booleans we can just do a == false/true and for others we might want to consider elvis operators or just if condition wrappers.

jesswrd replied:
Since we want an exact Groovy to Kotlin conversion, we will handle NPE handing during cleanup in a different PR.

ash-google commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTaskHelper.kt:
```diff
@@ -0,0 +1,121 @@
+
+    @VisibleForTesting
+    internal fun checkPreConditions() {
+        if (baseFlutterTask.sourceDir == null || !baseFlutterTask.sourceDir!!.isDirectory) {
```
I don't think the `!!` is required here since we're already ensuring it is not null.

jesswrd replied:
Since we want an exact Groovy to Kotlin conversion, we will handle NPE handing during cleanup in a different PR.

ash-google commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTaskHelper.kt:
```diff
@@ -0,0 +1,121 @@
+            baseFlutterTask.trackWidgetCreation?.let {
+                args("-dTrackWidgetCreation=${baseFlutterTask.trackWidgetCreation}")
+            }
+            baseFlutterTask.splitDebugInfo?.let {
+                args("-dSplitDebugInfo=${baseFlutterTask.splitDebugInfo}")
+            }
+            if (baseFlutterTask.treeShakeIcons == true) {
+                args("-dTreeShakeIcons=true")
+            }
+            if (baseFlutterTask.dartObfuscation == true) {
+                args("-dDartObfuscation=true")
+            }
+            baseFlutterTask.dartDefines?.let {
+                args("--DartDefines=${baseFlutterTask.dartDefines}")
+            }
+            baseFlutterTask.bundleSkSLPath?.let {
+                args("-dBundleSkSLPath=${baseFlutterTask.bundleSkSLPath}")
+            }
+            baseFlutterTask.codeSizeDirectory?.let {
+                args("-dCodeSizeDirectory=${baseFlutterTask.codeSizeDirectory}")
+            }
+            baseFlutterTask.flavor?.let {
+                args("-dFlavor=${baseFlutterTask.flavor}")
+            }
+            baseFlutterTask.extraGenSnapshotOptions?.let {
+                args("--ExtraGenSnapshotOptions=${baseFlutterTask.extraGenSnapshotOptions}")
+            }
+            baseFlutterTask.frontendServerStarterPath?.let {
+                args("-dFrontendServerStarterPath=${baseFlutterTask.frontendServerStarterPath}")
+            }
+            baseFlutterTask.extraFrontEndOptions?.let {
+                args("--ExtraFrontEndOptions=${baseFlutterTask.extraFrontEndOptions}")
+            }
```
Since we're not using the parameter from `.let`, let's just replace these with if null checks instead so it's a little more readable.

jesswrd committed:
added tests to baseFlutterTaskHelperTest and cleaned up

gmackall commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTask.kt:
```diff
@@ -0,0 +1,144 @@
+    fun buildBundle() {
+        val helper = BaseFlutterTaskHelper(baseFlutterTask = this)
+        helper.checkPreConditions()
+        intermediateDir!!.mkdirs()
```
I believe this was done because the original code behaved this way i.e., 
1. `intermediateDir` took a Java `File`, and so was nullable and 
2. we would throw a NPE on the old line 1642 of `flutter.groovy`, `intermediateDir` was null. 

I know we had initially spoken about not making behavior changes in the conversion. Do you think we should amend that for the case of null pointer exceptions? I think that would be reasonable, but we should be explicit.

reidbaker replied:
@gmackall talked with some members of the flutter android team and the decision was made to file an issue to track fixing this issue and keeping the code as it was before.
There is a balance between fixing issues as we see them and introducing new behavior and the goal of this project (language conversion + unit testing) is to minimize (to the point of zero) the introduction of new behavior if possible. 

Keeping an NPE we know about hurts but we are following our own principals in this project. After the conversion is done we do intend to fix many of the issues we have found.

jesswrd committed:
cleanup and safe state

jesswrd committed:
cleanup and add intermediateDir.mkdirs() into checkPreConditons

jesswrd committed:
Merge branch 'master' into i162111-convert-BaseFlutterTask

jesswrd committed:
Merge remote-tracking branch 'jesswrd/i162111-convert-BaseFlutterTask' into i162111-convert-BaseFlutterTask

jesswrd committed:
same as groovy no NPE handling

addressed comments

changed back to list

debugging windows integration test

ktlint formatting matches CI

delete commas and added platform-specific path separators

update comment

updated comment and removed debugging commands

cut down the mocked function calls

jesswrd commented on packages/flutter_tools/gradle/src/test/kotlin/BaseFlutterTaskHelperTest.kt:
```diff
@@ -0,0 +1,745 @@
+import kotlin.test.assertTrue
+
+class BaseFlutterTaskHelperTest {
+    object BaseFlutterTaskPropertiesTest {
```
Best practice for Kotlin when sharing constants, values, and/or variables across files.

jesswrd commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTaskHelper.kt:
```diff
@@ -0,0 +1,121 @@
+    // rather than the TargetPlatform. This allows multiple builds to share the same
+    // cache.
+    @VisibleForTesting
+    internal fun generateRuleNames(baseFlutterTask: BaseFlutterTask): List<String> {
```
done

jesswrd commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTaskHelper.kt:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @VisibleForTesting
+    internal fun createExecSpecActionFromTask(): Action<ExecSpec> =
```
done

jesswrd commented on packages/flutter_tools/gradle/src/test/kotlin/BaseFlutterTaskTest.kt:
```diff
@@ -0,0 +1,582 @@
+    private val extraGenSnapshotOptionsTest = "--debugger"
+    private val splitDebugInfoTest = "/path/to/build/debug_info_directory"
+    private val codeSizeDirectoryTest = "/path/to/build/code_size_directory"
+    private val minSDKVersionTest = 21
```
I'm inclined to not use that way of getting the minSDKVersion because I would have to create a mockProject to use the get minSDKVersion function [here](https://github.com/flutter/flutter/blob/b16430b2fd57a5aa6b6c680cdbda5582506fe120/packages/flutter_tools/gradle/src/main/kotlin/DependencyVersionChecker.kt#L171-L181). The mockProject would only be used to get the minSDKVersion. It looks like some functions and variables would have to be shared across files too.

reidbaker commented:
This is looking good. Super close to being able to land. 

Something I had a comment for but removed was making a utility class for mocking that let you pass in values. I think because you have really a test for null values and a test for non null it is probably not worth creating a utility class.

jesswrd committed:
Merge remote-tracking branch 'upstream/master' into i162111-convert-BaseFlutterTask

jesswrd committed:
removed baseFlutterTask

jesswrd commented on packages/flutter_tools/gradle/src/main/kotlin/BaseFlutterTask.kt:
```diff
@@ -0,0 +1,152 @@
+    var flavor: String? = null
+
+    /**
+     * Gets the dependency file(s) based on the path from the intermediate directory.
```
done

jesswrd commented on packages/flutter_tools/gradle/src/test/kotlin/BaseFlutterTaskHelperTest.kt:
```diff
@@ -0,0 +1,745 @@
+        every { mockExecSpec.args("--local-engine-src-path", localEngineSrcPath) } returns mockExecSpec
+
+        val localEngineHost = baseFlutterTask.localEngineHost
+        every { mockExecSpec.args("--local-engine-host", localEngineHost) } returns mockExecSpec
```
done
